### PR TITLE
refactor: Use polling model for workflow phase metric

### PR DESCRIPTION
**Pull request by simster7**
Fixes: https://github.com/argoproj/argo/issues/4551

Signed-off-by: Simon Behar <simbeh7@gmail.com>

Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

File: workflow/controller/controller.go [hunk 1]
```diff
@@ -54,10 +54,6 @@ import (
 	"github.com/argoproj/argo/workflow/util"
 )
 
-const enoughTimeForInformerSync = 1 * time.Second
-
-const semaphoreConfigIndexName = "bySemaphoreConfigMap"
-
```
Comment (simster7) [inline workflow/controller/controller.go:60]: Some pork-barrel style changes in this PR

File: workflow/controller/controller.go [hunk 2]
```diff
@@ -200,7 +198,9 @@ func (wfc *WorkflowController) Run(ctx context.Context, wfWorkers, podWorkers in
 
 	go wfc.runTTLController(ctx)
 	go wfc.runCronController(ctx)
+
 	go wfc.metrics.RunServer(ctx)
+	go wait.Until(wfc.syncWorkflowPhaseMetrics, 5*time.Second, ctx.Done())
```
Comment (simster7) [inline workflow/controller/controller.go:203]: I think 5 seconds is a good balance here

**Review by alexec — CHANGES_REQUESTED** [review]
Without @jessesuen to discus, can we set the polling interval to 1m. I'd rather have lower system load than up-to-date values. I will approve that.

**Comment by simster7** [comment]
> Without @jessesuen to discus, can we set the polling interval to 1m. I'd rather have lower system load than up-to-date values. I will approve that.

I'm okay waiting for @jessesuen to discuss. In my opinion 1m is too high, esp. since the load is minimal since we read from the informer

**Comment by alexec** [comment]
> since the load is minimal since we read from the informer 

Good point. 

This is a blocking issue for v2.12, so I suggest set it to 15s as a compromise and these discus with @jesse next week.

**Review by alexec — APPROVED** [review]

File: workflow/controller/indexes/labels.go [hunk 3]
```diff
@@ -3,14 +3,48 @@ package indexes
...
 	"k8s.io/client-go/tools/cache"
+
+	"github.com/argoproj/argo/workflow/util"
 )
 
 func MetaNamespaceLabelIndex(namespace, label string) string {
 	return namespace + "/" + label
 }
 
+func MetaLabelIndexFunc(label string) cache.IndexFunc {
```
[the remark this answers is not in the document]
Comment (simster7) [reply workflow/controller/indexes/labels.go:18]: Done

**Comment by alexec** [comment]
Merged into https://github.com/argoproj/argo/releases/tag/v2.12.0-rc4

**Pull request merged**